x-pack/plugins/osquery/server/usage/collector.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { CoreSetup, SavedObjectsClientContract } from '../../../../../src/core/server';
+import { CollectorFetchContext } from '../../../../../src/plugins/usage_collection/server';
+import { createMetricObjects } from '../routes/usage';
+import { getBeatUsage, getLiveQueryUsage, getPolicyLevelUsage } from './fetchers';
+import { CollectorDependencies, usageSchema, UsageData } from './types';
+
+export type RegisterCollector = (deps: CollectorDependencies) => void;
+export async function getInternalSavedObjectsClient(core: CoreSetup) {
+  return core.getStartServices().then(async ([coreStart]) => {
+    return coreStart.savedObjects.createInternalRepository();
+  });
+}
+
+export const registerCollector: RegisterCollector = ({ core, osqueryContext, usageCollection }) => {
+  if (!usageCollection) {
+    return;
+  }
+  const collector = usageCollection.makeUsageCollector<UsageData>({
+    type: 'osquery',
+    schema: usageSchema,
+    isReady: async () => {
+      const internalSavedObjectsClient = await getInternalSavedObjectsClient(core);
+      const savedObjectsClient = (internalSavedObjectsClient as unknown) as SavedObjectsClientContract;
+      return await createMetricObjects(savedObjectsClient);
+    },
+    fetch: async ({ esClient }: CollectorFetchContext): Promise<UsageData> => {
+      const internalSavedObjectsClient = await getInternalSavedObjectsClient(core);
+      const savedObjectsClient = (internalSavedObjectsClient as unknown) as SavedObjectsClientContract;
```

Comment:
I haven't seen code like this before. Why are you casting like this?

Reply:
honestly, i cribbed this from [here](https://github.com/elastic/kibana/blob/master/x-pack/plugins/security_solution/server/usage/collector.ts#L454). i didn't check to see if this could be more type safe since it didn't blow up and my attention was elsewhere; i'll give it a look and report back.

Reply:
++ this casting doesnt look right. You probably need `ISavedObjectsClient` or `InternalRepository`. havent pulled the code to double check but let me know if you get stuck

Reply:
from what I can tell, the code in `getInternalSavedObjectsClient` returns a repository that the `SavedObjectsClient` class uses for things to interface with. i ended up just constructing a new client with the resulting repository object.

the types are happy, and the code seems to be working, but i didn't check to see if there were any adverse effects of instantiating a client - potential perf or singleton issues come to mind.